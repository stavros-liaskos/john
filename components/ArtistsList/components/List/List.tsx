import LastFm from '../../../Icons/lastfm';
import Spotify from '../../../Icons/spotify';
import Button from '../../../Button/Button';
import React from 'react';
import { ArtistsListProp } from '../../ArtistsList';

const ICON_SIZE = 30;

const List = ({ i18n, artistsList, onButtonClick, artistLoading }: ArtistsListProp) => {
  if (!i18n || !i18n.btnTxt || !i18n.noArtists || !artistsList?.length || !onButtonClick) {
    return null;
  }

  return (
    <>
      {artistsList.map((artist, index: number) => (
        <div
          className="flex justify-between md:justify-center items-center dark:even:bg-gh-darkly even:bg-gray-100"
          key={artist.id}
        >
          <p className="grow text-clip rr-text">{artist.name}</p>
          <div className="flex basis-2 mx-4 md:mx-8">
            {artist.lastfmUri && (
              <a className="inline" href={artist.lastfmUri}>
                <LastFm width={ICON_SIZE} height={ICON_SIZE} />
              </a>
            )}
            {artist.spotifyUri && (
              <a className="inline" href={artist.spotifyUri}>
                <Spotify width={ICON_SIZE} height={ICON_SIZE} />
              </a>
            )}
          </div>
          <Button
            i18n={i18n.btnTxt}
            handleClick={onButtonClick}
            handleClickArg={artist.id}
            className={`btn-small lg:ml-8 my-2 ${index % 2 ? '!border-gh-dark' : ''}`}
            disabled={!!artistLoading && artist.id === artistLoading}
            loading={!!artistLoading && artist.id === artistLoading}
          />
        </div>
      ))}
    </>
  );
};

export default List;
