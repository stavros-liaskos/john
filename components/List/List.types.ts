export interface ListProps {
  i18n: { unfollow: string };
}
