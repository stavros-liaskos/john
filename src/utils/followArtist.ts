import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
