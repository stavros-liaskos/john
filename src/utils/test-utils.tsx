import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
