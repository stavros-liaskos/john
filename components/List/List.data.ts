import { ListEl } from './List.types';

export const listData: ListEl[] = [
  {
    followedByUser: false,
    lastfmUri: 'https://www.last.fm/music/Liedfett',
    name: 'Liedfett',
    spotifyUri: '',
  },
  {
    followedByUser: false,
    lastfmUri: 'https://www.last.fm/music/Leaders+of+the+New+School',
    name: 'Leaders of the New School',
    spotifyUri: '',
  },
  {
    followedByUser: false,
    lastfmUri: 'https://www.last.fm/music/lo+&+leduc',
    name: 'lo & leduc',
    spotifyUri: '',
  },
];

export const listI18n = {
  unfollow: 'unfollow',
};
