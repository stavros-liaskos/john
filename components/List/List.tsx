import React, { useState } from 'react';
import { ListProps, ListEl } from './List.types';
import Spotify from '../Icons/spotify';
import LastFm from '../Icons/lastfm';
import Button from '../Button/Button';

const ICON_SIZE = 30;

const List: React.FunctionComponent<ListProps> = ({ list, i18n }) => {
  const [artistLoading, setArtistLoading] = useState<string | null>(null);

  if (!list || !i18n || !i18n.unfollow) {
    return null;
  }

  const followArtist = (artistData: ListEl) => {
    setArtistLoading(artistData.name);
    fetch('/me/unfollow', {
      method: 'POST',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json',
      },
      referrerPolicy: 'no-referrer',
      body: JSON.stringify(artistData),
    })
      .then(() => {
        console.log(`${artistData.name} successfully unfollowed`);
      })
      .catch(error => {
        console.error('Error:', JSON.stringify(error));
      })
      .finally(() => {
        setTimeout(() => setArtistLoading(null), 2000);
      });
  };

  return (
    <div style={{ overflow: 'auto' }}>
      {' '}
      {/* TODO use tailwind */}
      {list.map((artist: ListEl, index: number) => (
        <div
          className="flex justify-between md:justify-center items-center dark:even:bg-gray-800 even:bg-gray-100"
          key={index}
        >
          <p className="grow text-clip dark:text-slate-300">{artist.name}</p>
          <div className="flex basis-2 mx-4 md:mx-8">
            {artist.lastfmUri && (
              <a className="inline" href={artist.lastfmUri}>
                <LastFm width={ICON_SIZE} height={ICON_SIZE} />
              </a>
            )}
            {artist.spotifyUri && (
              <a className="inline" href={artist.spotifyUri}>
                <Spotify width={ICON_SIZE} height={ICON_SIZE} />
              </a>
            )}
          </div>
          <Button
            i18n={i18n.unfollow}
            onClick={() => followArtist(artist)}
            className={`btn-small lg:ml-8 my-2 ${index % 2 ? '!border-zinc-900' : ''}`}
            disabled={!!artistLoading && artist.name === artistLoading}
            loading={!!artistLoading && artist.name === artistLoading}
          />
        </div>
      ))}
    </div>
  );
};

export default List;
