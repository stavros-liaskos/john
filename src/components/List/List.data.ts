import { ListEl } from './List.types';

export const listData: ListEl[] = [
  {
    name: 'Led Zeppeling',
    spotifyUrl: 'www.google.com',
  },
  {
    name: 'YOYOO',
    lastFmUrl: 'www.google.com',
  },
  {
    name: 'Art Beckley',
    lastFmUrl: 'www.google.com',
    spotifyUrl: 'www.google.com',
  },
];

export const listI18n = {
  unfollow: 'unfollow',
};
