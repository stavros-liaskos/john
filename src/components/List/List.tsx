import React from 'react';
import { ListProps, ListEl } from './List.types';
import Spotify from '../Icons/spotify';
import LastFm from '../Icons/lastfm';

const List: React.FunctionComponent<ListProps> = ({ list, i18n }) => {
  if (!list || !i18n || !i18n.unfollow) {
    return null;
  }

  return (
    <div className="border border-indigo-100">
      {list.map((el: ListEl, index: number) => (
        <div className="flex justify-between items-center px-8 even:bg-gray-100" key={index}>
          <p>{el.name}</p>
          <div className="flex">
            {el.lastFmUrl && (
              <a className="inline" href={el.lastFmUrl}>
                <LastFm />
              </a>
            )}
            {el.spotifyUrl && (
              <a className="inline" href={el.spotifyUrl}>
                <Spotify />
              </a>
            )}
          </div>
          <button className="py-2 px-3 bg-indigo-800 text-white text-sm font-semibold rounded-md shadow focus:outline-none">
            {i18n.unfollow}
          </button>
        </div>
      ))}
    </div>
  );
};

export default List;
