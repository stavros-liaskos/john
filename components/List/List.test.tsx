import React from 'react';
import { render } from '@testing-library/react';
import List from './List';
import { listI18n } from './List.data';
import { beforeEachTest } from '../../utils/test-utils';
import followedArtists from '../../mocks/responses/followed-artists.json';

describe('List', () => {
  beforeEach(() => {
    beforeEachTest();
    jest
      .spyOn(global, 'fetch')
      .mockResolvedValue(new Response(JSON.stringify({ json: followedArtists }), { status: 200, statusText: 'OK' }));
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<List />);
  });

  it('unfollows artist on "unfollow" btn click', () => {
    const { container } = render(<List i18n={listI18n} />);

    expect(container).toMatchSnapshot();

    // const unfollowBtns = getAllByText(container, 'unfollow');
    // expect(unfollowBtns.length).toEqual(2);
  });
});
