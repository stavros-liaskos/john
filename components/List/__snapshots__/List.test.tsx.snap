// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List unfollows artist on "unfollow" btn click 1`] = `<div />`;
